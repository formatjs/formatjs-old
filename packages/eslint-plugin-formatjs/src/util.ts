import {TSESTree, AST_NODE_TYPES} from '@typescript-eslint/typescript-estree';
import {Scope} from 'eslint';

export interface MessageDescriptor {
  id?: string;
  defaultMessage?: string;
  description?: string;
}

export interface MessageDescriptorNodeInfo {
  message: MessageDescriptor;
  messageNode?: TSESTree.Property['value'] | TSESTree.JSXAttribute['value'];
  descriptionNode?: TSESTree.Property['value'] | TSESTree.JSXAttribute['value'];
}

function isStringLiteral(node: TSESTree.Node): node is TSESTree.StringLiteral {
  return node.type === AST_NODE_TYPES.Literal && typeof node.value === 'string';
}

function findReferenceImport(
  id: TSESTree.Identifier,
  importedVars: Scope.Variable[]
) {
  return importedVars.find(
    v => !!v.references.find(ref => ref.identifier === id)
  );
}

function isIntlFormatMessageCall(node: TSESTree.Node) {
  return (
    node.type === AST_NODE_TYPES.CallExpression &&
    node.callee.type === AST_NODE_TYPES.MemberExpression &&
    node.callee.object.type === AST_NODE_TYPES.Identifier &&
    node.callee.object.name === 'intl' &&
    node.callee.property.type === AST_NODE_TYPES.Identifier &&
    node.callee.property.name === 'formatMessage' &&
    node.arguments.length >= 1 &&
    node.arguments[0].type === AST_NODE_TYPES.ObjectExpression
  );
}

function isSingleMessageDescriptorDeclaration(
  id: TSESTree.LeftHandSideExpression,
  importedVars: Scope.Variable[]
) {
  if (id.type !== AST_NODE_TYPES.Identifier) {
    return false;
  }
  const importedVar = findReferenceImport(id, importedVars);
  if (!importedVar) {
    return false;
  }
  return importedVar.name === '_' || importedVar.name === 'defineMessage';
}
function isMultipleMessageDescriptorDeclaration(
  id: TSESTree.LeftHandSideExpression,
  importedVars: Scope.Variable[]
) {
  if (id.type !== AST_NODE_TYPES.Identifier) {
    return false;
  }
  const importedVar = findReferenceImport(id, importedVars);
  if (!importedVar) {
    return false;
  }
  return importedVar.name === 'defineMessages';
}

function extractMessageDescriptor(
  node?: TSESTree.Expression
): MessageDescriptorNodeInfo | undefined {
  if (!node || node.type !== AST_NODE_TYPES.ObjectExpression) {
    return;
  }
  const result: MessageDescriptorNodeInfo = {
    message: {},
    messageNode: undefined,
    descriptionNode: undefined,
  };
  for (const prop of node.properties) {
    if (
      prop.type !== AST_NODE_TYPES.Property ||
      prop.key.type !== AST_NODE_TYPES.Identifier
    ) {
      continue;
    }
    const value = isStringLiteral(prop.value) ? prop.value.value : undefined;
    switch (prop.key.name) {
      case 'defaultMessage':
        result.messageNode = prop.value;
        result.message.defaultMessage = value;
        break;
      case 'description':
        result.descriptionNode = prop.value;
        result.message.description = value;
        break;
      case 'id':
        result.message.id = value;
        break;
    }
  }
  return result;
}

function extractMessageDescriptorFromJSXElement(
  node?: TSESTree.JSXOpeningElement
):
  | [MessageDescriptorNodeInfo, TSESTree.ObjectExpression | undefined]
  | undefined {
  if (!node || !node.attributes) {
    return;
  }
  let values: TSESTree.ObjectExpression | undefined;
  const result: MessageDescriptorNodeInfo = {
    message: {},
    messageNode: undefined,
    descriptionNode: undefined,
  };
  for (const prop of node.attributes) {
    if (
      prop.type !== AST_NODE_TYPES.JSXAttribute ||
      prop.name.type !== AST_NODE_TYPES.JSXIdentifier
    ) {
      continue;
    }
    const key = prop.name;
    switch (key.name) {
      case 'defaultMessage':
        result.messageNode = prop.value;
        if (
          prop.value?.type === AST_NODE_TYPES.Literal &&
          typeof prop.value.value === 'string'
        ) {
          result.message.defaultMessage = prop.value.value;
        }
        break;
      case 'description':
        result.descriptionNode = prop.value;
        if (
          prop.value?.type === AST_NODE_TYPES.Literal &&
          typeof prop.value.value === 'string'
        ) {
          result.message.description = prop.value.value;
        }
        break;
      case 'id':
        if (
          prop.value?.type === AST_NODE_TYPES.Literal &&
          typeof prop.value.value === 'string'
        ) {
          result.message.id = prop.value.value;
        }
        break;
      case 'values':
        if (
          prop.value?.type === AST_NODE_TYPES.JSXExpressionContainer &&
          prop.value.expression.type === AST_NODE_TYPES.ObjectExpression
        ) {
          values = prop.value.expression;
        }
        break;
    }
  }
  if (!Object.keys(result.message).length) {
    return;
  }
  return [result, values];
}

function extractMessageDescriptors(node?: TSESTree.Expression) {
  if (
    !node ||
    node.type !== AST_NODE_TYPES.ObjectExpression ||
    !node.properties.length
  ) {
    return [];
  }
  const msgs = [];
  for (const prop of node.properties) {
    if (prop.type !== AST_NODE_TYPES.Property) {
      continue;
    }
    const msg = prop.value;
    if (msg.type !== AST_NODE_TYPES.ObjectExpression) {
      continue;
    }
    const nodeInfo = extractMessageDescriptor(msg);
    if (nodeInfo) {
      msgs.push(nodeInfo);
    }
  }
  return msgs;
}

export function extractMessages(
  node: TSESTree.Node,
  importedMacroVars: Scope.Variable[],
  excludeMessageDeclCalls = false
): Array<[MessageDescriptorNodeInfo, TSESTree.Expression | undefined]> {
  if (node.type === AST_NODE_TYPES.CallExpression) {
    const expr = node;
    const fnId = expr.callee;
    if (
      (!excludeMessageDeclCalls &&
        isSingleMessageDescriptorDeclaration(fnId, importedMacroVars)) ||
      isIntlFormatMessageCall(node)
    ) {
      const msgDescriptorNodeInfo = extractMessageDescriptor(expr.arguments[0]);
      if (msgDescriptorNodeInfo) {
        return [[msgDescriptorNodeInfo, expr.arguments[1]]];
      }
    } else if (
      !excludeMessageDeclCalls &&
      isMultipleMessageDescriptorDeclaration(fnId, importedMacroVars)
    ) {
      return extractMessageDescriptors(expr.arguments[0]).map(msg => [
        msg,
        undefined,
      ]);
    }
  } else if (
    node.type === AST_NODE_TYPES.JSXOpeningElement &&
    node.name &&
    node.name.type === AST_NODE_TYPES.JSXIdentifier &&
    node.name.name === 'FormattedMessage'
  ) {
    const msgDescriptorNodeInfo = extractMessageDescriptorFromJSXElement(node);
    if (msgDescriptorNodeInfo) {
      return [msgDescriptorNodeInfo];
    }
  }
  return [];
}
