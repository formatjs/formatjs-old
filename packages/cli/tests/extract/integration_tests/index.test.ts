import path from 'path';
import util from 'util';
import childProcess from 'child_process';
import {readJSON, mkdirp, readdir} from 'fs-extra';
import _rimraf from 'rimraf';
const exec = util.promisify(childProcess.exec);
const rimraf = util.promisify(_rimraf);

const BIN_PATH = path.resolve(__dirname, '../../../bin/formatjs');
const ARTIFACT_PATH = path.resolve(__dirname, 'test_artifacts');

beforeEach(async () => {
  await mkdirp(path.join(__dirname, 'test_artifacts'));
  await rimraf(ARTIFACT_PATH);
});

test('basic case: help', async () => {
  const {stdout, stderr} = await exec(`${BIN_PATH} extract --help`);
  expect(stdout).toMatchSnapshot();
  expect(stderr).toBe('');
}, 10000);

test('basic case: defineMessages -> stdout', async () => {
  const {stdout, stderr} = await exec(
    `${BIN_PATH} extract ${path.join(__dirname, 'defineMessages/actual.js')}`
  );
  expect(JSON.parse(stdout)).toMatchSnapshot();
  expect(stderr).toBe('');
}, 10000);

test('[glob] basic case: defineMessages -> stdout', async () => {
  const {stdout, stderr} = await exec(
    `${BIN_PATH} extract ${path.join(__dirname, 'defineMessages/*.js')}`
  );
  expect(JSON.parse(stdout)).toMatchSnapshot();
  expect(stderr).toBe('');
}, 10000);

test('basic case: defineMessages -> directory', async () => {
  process.chdir(__dirname);
  const {stdout, stderr} = await exec(
    `${BIN_PATH} extract defineMessages/actual.js --messages-dir ${ARTIFACT_PATH}`
  );
  expect(stdout).toBe('');
  expect(stderr).toBe('');
  // Write to test_artifacts/defineMessages/actual.json
  expect(
    await readdir(path.join(ARTIFACT_PATH, 'defineMessages'))
  ).toMatchSnapshot();
  expect(
    await readJSON(path.join(ARTIFACT_PATH, 'defineMessages/actual.json'))
  ).toMatchSnapshot();
}, 10000);

test('typescript -> stdout', async () => {
  const {stdout, stderr} = await exec(
    `${BIN_PATH} extract ${path.join(__dirname, 'typescript/actual.tsx')}`
  );
  expect(JSON.parse(stdout)).toMatchSnapshot();
  expect(stderr).toBe('');
}, 10000);

test('ignore -> stdout', async () => {
  const ignore = "--ignore '*.ignore.*'";

  const tsResult = await exec(
    `${BIN_PATH} extract ${path.join(
      __dirname,
      'typescript/actual.tsx'
    )} ${ignore}`
  );
  expect(JSON.parse(tsResult.stdout)).toMatchSnapshot();
  expect(tsResult.stderr).toBe('');

  const jsResult = await exec(
    `${BIN_PATH} extract ${path.join(
      __dirname,
      'defineMessages/*.js'
    )} ${ignore}`
  );
  expect(JSON.parse(jsResult.stdout)).toMatchSnapshot();
  expect(jsResult.stderr).toBe('');
}, 10000);
