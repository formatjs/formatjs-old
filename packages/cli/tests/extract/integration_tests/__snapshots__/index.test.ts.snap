// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[glob] basic case: defineMessages -> stdout 1`] = `
Array [
  Object {
    "defaultMessage": "ignore",
    "id": "ignore",
  },
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
    "id": "app.home.kittens",
  },
  Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
    "id": "trailing.ws",
  },
  Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
    "id": "escaped.apostrophe",
  },
  Object {
    "defaultMessage": "No ID",
    "description": "no ID",
    "id": "ae494",
  },
]
`;

exports[`basic case: defineMessages -> directory 1`] = `
Array [
  "actual.json",
]
`;

exports[`basic case: defineMessages -> directory 2`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
    "id": "app.home.kittens",
  },
  Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
    "id": "trailing.ws",
  },
  Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
    "id": "escaped.apostrophe",
  },
  Object {
    "defaultMessage": "No ID",
    "description": "no ID",
    "id": "ae494",
  },
]
`;

exports[`basic case: defineMessages -> stdout 1`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
    "id": "app.home.kittens",
  },
  Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
    "id": "trailing.ws",
  },
  Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
    "id": "escaped.apostrophe",
  },
  Object {
    "defaultMessage": "No ID",
    "description": "no ID",
    "id": "ae494",
  },
]
`;

exports[`basic case: help 1`] = `
"Usage: formatjs extract [options] [files...]

Extract string messages from React components that use react-intl.
The input language is expected to be TypeScript or ES2017 with JSX.

Options:
  --messages-dir <dir>                                  The target location where the plugin will output a \`.json\` file corresponding to each component from which React Intl messages were extracted. If not provided, the extracted message descriptors will be printed to standard output.
  --out-file <path>                                     The target file path where the plugin will output an aggregated \`.json\` file of allthe translations from the \`files\` supplied.
  This flag will ignore --messages-dir
  --id-interpolation-pattern <pattern>                  If certain message descriptors don't have id, this \`pattern\` will be used to automaticallygenerate IDs for them. Default to \`[contenthash:5]\`.
  See https://github.com/webpack/loader-utils#interpolatename for sample patterns
  --extract-source-location                             Whether the metadata about the location of the message in the source file should be extracted. If \`true\`, then \`file\`, \`start\`, and \`end\` fields will exist for each extracted message descriptors. (default: false)
  --module-source-name <name>                           The ES6 module source name of the React Intl package. Defaults to: \`\\"react-intl\\"\`, but can be changed to another name/path to React Intl.
  --remove-default-message                              Remove \`defaultMessage\` field in generated js after extraction (default: false)
  --additional-component-names <comma-separated-names>  Additional component names to extract messages from, e.g: \`['FormattedFooBarMessage']\`. **NOTE**: By default we check for the fact that \`FormattedMessage\` & \`FormattedHTMLMessage\` are imported from \`moduleSourceName\` to make sure variable alias works. This option does not do that so it's less safe.
  --extract-from-format-message-call                    Opt-in to extract from \`intl.formatMessage\` call with the same restrictions, e.g: has to be called with object literal such as \`intl.formatMessage({ id: 'foo', defaultMessage: 'bar', description: 'baz'})\` (default: false)
  --ignore <files>                                      List of glob paths to **not** extract translations from.
  -h, --help                                            output usage information
"
`;

exports[`ignore -> stdout 1`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
    "id": "app.home.kittens",
  },
  Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
    "id": "trailing.ws",
  },
  Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
    "id": "escaped.apostrophe",
  },
  Object {
    "defaultMessage": "No ID",
    "description": "no ID",
    "id": "ae494",
  },
  Object {
    "defaultMessage": "formatted message",
    "description": "foo",
    "id": "inline",
  },
]
`;

exports[`ignore -> stdout 2`] = `
Array [
  Object {
    "defaultMessage": "ignore",
    "id": "ignore",
  },
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
    "id": "app.home.kittens",
  },
  Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
    "id": "trailing.ws",
  },
  Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
    "id": "escaped.apostrophe",
  },
  Object {
    "defaultMessage": "No ID",
    "description": "no ID",
    "id": "ae494",
  },
]
`;

exports[`typescript -> stdout 1`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
    "id": "app.home.kittens",
  },
  Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
    "id": "trailing.ws",
  },
  Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
    "id": "escaped.apostrophe",
  },
  Object {
    "defaultMessage": "No ID",
    "description": "no ID",
    "id": "ae494",
  },
  Object {
    "defaultMessage": "formatted message",
    "description": "foo",
    "id": "inline",
  },
]
`;
